mule-retry-api/src/main/java/org/mule/runtime/retry/api/policy/RetryPolicyTemplate.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright 2023 Salesforce, Inc. All rights reserved.
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.runtime.retry.api.policy;
+
+import static org.mule.runtime.retry.api.policy.util.rx.ImmediateScheduler.IMMEDIATE_SCHEDULER;
+
+import static java.util.function.Function.identity;
+
+import org.mule.api.annotation.NoImplement;
+import org.mule.runtime.api.scheduler.Scheduler;
+import org.mule.runtime.retry.api.RetryCallback;
+import org.mule.runtime.retry.api.RetryContext;
+import org.mule.runtime.retry.api.RetryNotifier;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+import org.reactivestreams.Publisher;
+
+/**
+ * A RetryPolicyTemplate creates a new {@link RetryPolicy} instance each time the retry goes into effect, thereby resetting any
+ * state the policy may have (counters, etc.)
+ * <p>
+ * A {@link RetryNotifier} may be set in order to take action upon each retry attempt.
+ */
+@NoImplement
+public interface RetryPolicyTemplate {
+
+  RetryPolicy createRetryInstance();
+
+  Map<Object, Object> getMetaInfo();
+
+  void setMetaInfo(Map<Object, Object> metaInfo);
+
+  RetryNotifier getNotifier();
+
+  void setNotifier(RetryNotifier retryNotifier);
+
+  /**
+   * Applies the retry policy by performing a blocking action.
+   *
+   * @param callback    a callback with the logic to be executed on each retry
+   * @param workManager the executor on which the retry operations are to be executed
+   * @return a {@link RetryContext}
+   */
+  RetryContext execute(RetryCallback callback, Executor workManager) throws Exception;
+
+  /**
+   * Applies the retry policy in a non blocking manner by transforming the given {@code publisher} into one configured to apply
+   * the retry logic.
+   *
+   * @param publisher a publisher with the items which might fail
+   * @param <T>       the generic type of the publisher's content
+   * @return a {@link Publisher} configured with the retry policy.
+   * @since 4.0
+   * @deprecated Use {@link #applyPolicy(Publisher, Scheduler)} instead
+   */
+  @Deprecated
+  default <T> Publisher<T> applyPolicy(Publisher<T> publisher) {
```

Comment:
this method uses reactivestreams, it should not be part of the api. Isn't this used only by the container?